File: app/src/main/java/com/example/test/di/MultiViewModelFactory.java
package com.example.test.di;

import java.util.Map;
import java.util.Set;

import javax.inject.Inject;
import javax.inject.Provider;

import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

public class MultiViewModelFactory implements ViewModelProvider.Factory {

    private final Map<Class<? extends ViewModel>, Provider<ViewModel>> viewModelFactories;

    @Inject
    public MultiViewModelFactory(Map<Class<? extends ViewModel>, Provider<ViewModel>> viewModelFactories) {
        this.viewModelFactories = viewModelFactories;
    }

    public Set<Class<? extends ViewModel>> getViewModelClasses() {
        return viewModelFactories.keySet();
    }

    @Override
    public <T extends ViewModel> T create(Class<T> modelClass) {
        Provider<ViewModel> provider = viewModelFactories.get(modelClass);
        if (provider == null) {
            throw new IllegalArgumentException("Unknown ViewModel class: " + modelClass);
        }
        return (T) provider.get();
    }

}

File: app/src/main/java/com/example/test/di/MultiViewModelFactory.kt
package com.example.test.di

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import javax.inject.Inject
import javax.inject.Provider

class MultiViewModelFactory @Inject constructor(private val viewModelFactories: Map<Class<out ViewModel>,
    @JvmSuppressWildcards Provider<ViewModel>>) : ViewModelProvider.Factory {

    fun getViewModelClasses(): Set<Class<out ViewModel>> {
        return viewModelFactories.keys
    }

    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        val provider = viewModelFactories[modelClass]
            ?: throw IllegalArgumentException("Unknown ViewModel class: + $modelClass")
        return provider.get() as T
    }
}